### ISSUE TITLE
Question: Correct `this` context when using `{{action}}` on a microstate.

### ISSUE THREAD

**Issue by brandynbennett** [microstates/ember#78]
How are you able to have the `this` keyword reference the Microstate instead of the component when calling a transition on a custom type using `{{action}}` (e.g. `{{action person.changeIdentity "Bart" "Simpson"}}`)? Whenever I try to invoke a transition directly on a microstate via `{{action}}`, `this` always refers to the component that had the `{{action}}` not the microstate object.

**Comment by brandynbennett** [comment]
When this addon update gets released what would I do differently to not use the `bind` helper? Seems like you can either make a microstate using `{{let}}` or with the `state` macro. Would those have the same problem?

**Comment by cowboyd** [comment]
@brandynbennett If you're using Microstates directly, you might want to use a `Store`. This will accomplish too things:

1) it binds `this` for you
2) the actual function is stable, meaning that it will not trigger re-renders. i.e. `person.changeIdentity` is the same function no matter what

**Comment by taras** [comment]
> To clarify I'm not using the ember-microstates addon.

You need to wrap a microstate in a Store, which creates closures from methods on the microstate. It does a bunch of other good things for you.

**Comment by taras** [comment]
@brandynbennett I'm going to release 0.12.0 so you can use the code in this repo.

**Comment by brandynbennett** [comment]
So I'm using the Store, but i'm still having the issue. I'm yielding the Store created microstate from a parent component and a child component is calling the transition. Working on putting up a sample repo to reproduce.

Here it is so far https://github.com/brandynbennett/microstates-test. Haven't reproduced the `this` issue yet but I'm running up into 

```
Uncaught TypeError: Class constructor Person cannot be invoked without 'new'
    at Microstate._class (microstates.es.js:1202)
    at new Microstate<Person> (microstates.es.js:1263)
    at create (microstates.es.js:1240)
    at Class.init (component.js:13)
    at Class.superWrapper [as init] (ember-utils.js:308)
    at Class.CoreObject (core_object.js:145)
    at Class.EmberObject (object.js:28)
    at Class.FrameworkObject (object.js:72)
    at Class (core_object.js:207)
    at Class (core_object.js:207)
```
 
Coming from:
```
function _class() {
      _classCallCheck(this, _class);

      return _possibleConstructorReturn(this, _getPrototypeOf(_class).apply(this, arguments));
    }
```

Happening here https://github.com/brandynbennett/microstates-test/blob/master/app/components/parent/component.js#L12

**Comment by taras** [comment]
@brandynbennett I just published `@microstates/ember`. You can install that package and use `state` helper. It'll take care of everything for you.

Here are the demos https://ember-microstates.netlify.com

**Comment by brandynbennett** [comment]
Awesome. I'll give that a shot. Thanks.

**Comment by taras** [comment]
@brandynbennett pls report how it goes.

**Comment by brandynbennett** [comment]
So I was able to reproduce what was happening in this example repo: https://github.com/brandynbennett/microstates-test

The issue ONLY happens when you are using a state class that extends another state class. If you `ember serve` that app you can change the `person` type to an `infant` type via a button. Once you do that and try clicking the "change identity" button again `this` is no longer the microstate but the `component`. Could it be possible that the `Store` is only putting a closure around the classes "own" properties and not that of the superclass?

**Comment by brandynbennett** [comment]
![kapture 2018-11-30 at 15 51 05](https://user-images.githubusercontent.com/4146629/49318936-c7053580-f4b7-11e8-9e91-a47a2f64b98f.gif)

**Comment by brandynbennett** [comment]
I simplified the example repo. I initialize the `Store` with an `Infant` type microstate which extends `Person` and trying to call `{{action person.changeIdentity}}` immediately fails.

**Comment by brandynbennett** [comment]
It happens regardless of whether I use the `state` helper or if I use the `Store` directly in the component.js

**Comment by taras** [comment]
> Could it be possible that the Store is only putting a closure around the classes "own" properties and not that of the superclass?

Yes, I believe that's what's happening here. Would you mind filling an issue with in Microstates.js repository?

**Comment by taras** [comment]
So, it looks like something else is failing because I'm seeing that it's creating closures around parent's transitions. https://runkit.com/taras/microsates---descendant-class-bug

I'm going to look at your example closer to figure out why your example is failing.

**Comment by taras** [comment]
@brandynbennett thank you for tracking this down. I'll get it fixed ASAP.

**Comment by brandynbennett** [comment]
@taras Thanks so much for looking into it. I really appreciate it. And thanks for making such an excellent, well-designed state management solution. It really does make state management fun.

**Comment by taras** [comment]
@brandynbennett I just published `@microstates/ember` vs 0.12.1. It should address your issue. Please try and let us know if it doesn't.

**Comment by brandynbennett** [comment]
Thanks so much! I’ll give it a shot.
